src/inference/src/dev/make_tensor.cpp

```diff
@@ -178,28 +180,63 @@ class AllocatedTensor : public ViewTensor {
                      shape,
                      [&] {
                          OPENVINO_ASSERT(allocator, "Allocator was not initialized");
-                         return const_cast<Allocator&>(allocator).allocate(element_type.size() * shape_size(shape));
+                         auto num_elements = shape_size(shape);
+                         auto data = const_cast<Allocator&>(allocator).allocate(element_type.size() * num_elements);
+                         init(data, element_type, shape);
+                         return data;
                      }()},
           m_allocator{allocator} {}
 
     ~AllocatedTensor() {
+        auto num_elements = get_size();
+        destroy(0, num_elements);
         m_allocator.deallocate(m_ptr, get_byte_size());
     }
 
     void set_shape(ov::Shape new_shape) override {
         if (m_shape == new_shape)
             return;
+
+        auto old_num_elements = get_size();
         auto old_byte_size = get_byte_size();
         m_shape = std::move(new_shape);
+        auto new_num_elements = get_size();
+
         if (get_byte_size() > old_byte_size) {
+            // allocate buffer and initialize objects from scratch
+            destroy(0, old_num_elements);
             m_allocator.deallocate(m_ptr, old_byte_size);
             m_ptr = m_allocator.allocate(get_byte_size());
+            init(m_ptr, m_element_type, m_shape);
+        } else {
+            // destroy only not needed objects
+            destroy(new_num_elements, old_num_elements);
```

Comment:
```
ov::Tensor t(string, {4, 4}); // all strings are initialized
t.set_shape({2, 2}); // some strings are deinitialized
t.set_shape(4, 4); // m_capacity is 16, no allocations are required. But in current logic we we destroy memory
```
Looks like we need to compare with `m_capacity` when decide about memory re-allocation

Reply:
This is a general problem of OV that it cannot use efficiently pre-allocated memory. This is not related only string type. I think we need to improve this in the scope of separate activity